src/compiler/checker.ts

```diff
@@ -6018,6 +6022,19 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return setTextRangeWorker(setOriginalNode(range, location), location);
         }
 
+        /**
+         * Same as expressionOrTypeToTypeNodeHelper, but also checks if the expression can be syntactically typed.
+         */
+        function expressionOrTypeToTypeNodeHelper(context: NodeBuilderContext, expr: Expression | JsxAttributeValue | undefined, type: Type, addUndefined?: boolean) {
+            const oldFlags = context.flags;
+            if (expr && !(context.flags & NodeBuilderFlags.NoSyntacticPrinter)) {
+                syntacticNodeBuilder.serializeTypeOfExpression(expr, context, addUndefined);
+            }
+            context.flags |= NodeBuilderFlags.NoSyntacticPrinter;
+            const result = expressionOrTypeToTypeNode(context, expr, type, addUndefined);

```

Comment:
`expressionOrTypeToTypeNodeHelper` (after rename)  may call `typeToTypeNodeHelper` which in turn will call `serializeTypeForDeclaration` and `serializeReturnTypeForSignature` and these would perform the check again. I can refactor those into `serializeTypeForDeclarationHelper` and `serializeReturnTypeForSignatureHelper`  which do not perform the check and `serializeTypeForDeclaration` and `serializeReturnTypeForSignature`  which would only be called from the top level. This would ensure the code does not check multiple times the same expression, and would remove the need for the `NoSyntacticPrinter` flag. 

The flag will disappear in the next PRs. It's definitely not something that will remain. When the syntactic builder actually builds the types and falls back to the checked builder when it can't the problem of duplicate checks will go away.

Should I do the  `serializeTypeForDeclarationHelper` and `serializeReturnTypeForSignatureHelper` refactor for now? Or ca we live with this until after the beta?